src/LtiAssignmentsGradesService.php

```diff
@@ -14,24 +14,48 @@ public function getScope(): array
         return $this->getServiceData()['scope'];
     }
 
+    // https://www.imsglobal.org/spec/lti-ags/v2p0#assignment-and-grade-service-claim
+    // When an LTI message is launching a resource associated to one and only one lineitem,
+    // the claim must include the endpoint URL for accessing the associated line item;
+    // in all other cases, this property must be either blank or not included in the claim.
+    public function getResourceLaunchLineItem(): ?LtiLineitem
+    {
+        $serviceData = $this->getServiceData();
+        if (empty($serviceData['lineitem'])) {
+            return null;
+        }
+
+        $lineitem = LtiLineitem::new()->setId($serviceData['lineitem']);
+        return $lineitem;
+    }
+
     public function putGrade(LtiGrade $grade, LtiLineitem $lineitem = null)
     {
         if (!in_array(LtiConstants::AGS_SCOPE_SCORE, $this->getScope())) {
             throw new LtiException('Missing required scope', 1);
         }
-        if ($lineitem !== null && empty($lineitem->getId())) {
-            $lineitem = $this->findOrCreateLineitem($lineitem);
-            $scoreUrl = $lineitem->getId();
-        } elseif ($lineitem === null && !empty($this->getServiceData()['lineitem'])) {
-            $scoreUrl = $this->getServiceData()['lineitem'];
-        } else {
-            $lineitem = LtiLineitem::new()
-                ->setLabel('default')
-                ->setScoreMaximum(100);
-            $lineitem = $this->findOrCreateLineitem($lineitem);
-            $scoreUrl = $lineitem->getId();
+
+        // If the line item already contains the Id, we do not need to find or create it.
+        if ($lineitem !== null) {
+            if (empty($lineitem->getId())) {
+                $lineitem = $this->findOrCreateLineitem($lineitem);
+            }
+        }
+
+        // Otherwise, if no line item is passed in, attempt to use the one associated with
+        // this launch. If none exists, create a default line item.
+        else {
+            $lineitem = $this->getResourceLaunchLineItem();
+            if (!$lineitem || empty($lineitem->getId())) {
```

Comment:
```suggestion
            if (!isset($lineitem, $lineitem->getId()))
```

Reply:
this seems fine if it's clearer for you -- to me it's a bit more difficult to parse, but that is just personal preference.